File: frontend/src/lib/auth/permissions.ts
import { useAppSelector } from '@/lib/store/store';

/** Permission codes — mirror of backend Permissions.cs / V005__auth.sql. */
export const P = {
  CustomerRead: 'customer.read',
  CustomerWrite: 'customer.write',
  VehicleRead: 'vehicle.read',
  VehicleWrite: 'vehicle.write',
  LookupRead: 'lookup.read',
  LookupManage: 'lookup.manage',
  QuotationRead: 'quotation.read',
  QuotationWrite: 'quotation.write',
  PolicyRead: 'policy.read',
  PolicyIssue: 'policy.issue',
  PolicyActivate: 'policy.activate',
  PolicyCancel: 'policy.cancel',
  PolicyRenew: 'policy.renew',
  ClaimRead: 'claim.read',
  ClaimFile: 'claim.file',
  ClaimReview: 'claim.review',
  ClaimApprove: 'claim.approve',
  ClaimReject: 'claim.reject',
  PaymentRead: 'payment.read',
  PaymentSettle: 'payment.settle',
  DashboardRead: 'dashboard.read',
} as const;

/** Minimum permission required to *view* each page. Mutations are gated server-side. */
export const ROUTE_PERMISSION: Record<string, string> = {
  '/': P.DashboardRead,
  '/customers': P.CustomerRead,
  '/vehicles': P.VehicleRead,
  '/quotations': P.QuotationRead,
  '/policies': P.PolicyRead,
  '/claims': P.ClaimRead,
  '/payments': P.PaymentRead,
  '/master': P.LookupManage,
};

/** Resolve the permission guarding a given pathname (longest matching prefix). */
export function requiredPermission(pathname: string): string | undefined {
  if (pathname === '/') return ROUTE_PERMISSION['/'];
  const match = Object.entries(ROUTE_PERMISSION)
    .filter(([path]) => path !== '/' && pathname.startsWith(path))
    .sort((a, b) => b[0].length - a[0].length)[0];
  return match?.[1];
}

export function usePermissions(): string[] {
  return useAppSelector((s) => s.auth.user?.permissions ?? []);
}

/** Returns true when no permission is required or the user holds it. */
export function useHasPermission(permission?: string): boolean {
  const permissions = usePermissions();
  return !permission || permissions.includes(permission);
}

File: backend/src/MotorInsurance.Application/Common/Authorization/Permissions.cs
namespace MotorInsurance.Application.Common.Authorization;

/// <summary>
/// Permission code constants. MUST match the seed values in db/scripts/V005__auth.sql
/// (the permission.code column) and the frontend mirror in lib/auth/permissions.ts.
/// </summary>
public static class Permissions
{
    public const string CustomerRead = "customer.read";
    public const string CustomerWrite = "customer.write";

    public const string VehicleRead = "vehicle.read";
    public const string VehicleWrite = "vehicle.write";

    public const string LookupRead = "lookup.read";
    public const string LookupManage = "lookup.manage";

    public const string QuotationRead = "quotation.read";
    public const string QuotationWrite = "quotation.write";

    public const string PolicyRead = "policy.read";
    public const string PolicyIssue = "policy.issue";
    public const string PolicyActivate = "policy.activate";
    public const string PolicyCancel = "policy.cancel";
    public const string PolicyRenew = "policy.renew";
    public const string PolicyEndorse = "policy.endorse";

    public const string ClaimRead = "claim.read";
    public const string ClaimFile = "claim.file";
    public const string ClaimReview = "claim.review";
    public const string ClaimApprove = "claim.approve";
    public const string ClaimReject = "claim.reject";

    public const string PaymentRead = "payment.read";
    public const string PaymentSettle = "payment.settle";

    public const string DashboardRead = "dashboard.read";

    public const string NotificationRead = "notification.read";
    public const string NotificationManage = "notification.manage";

    public const string RatingRead = "rating.read";
    public const string RatingManage = "rating.manage";

    public const string UserRead = "user.read";
    public const string UserManage = "user.manage";
}
